src/plugins/discourse/weights.js

```diff
@@ -87,3 +104,30 @@ export function _trustLevelWeight({trustLevel}: User): NodeWeight {
   }
   return weight;
 }
+
+export function _categoryWeight(
+  category: DiscourseCategoryId,
+  {categoryWeights = new Map(), defaultCategoryWeight = 1}: WeightsConfig
+): NodeWeight {
+  const categoryString = category.toString();
+  const weight = categoryWeights.get(categoryString);
+  if (weight != null) return weight;
+
+  return defaultCategoryWeight;
+}
+
+export function _weightFromTags(
+  tags: $ReadOnlyArray<Tag>,
+  {tagWeights, defaultTagWeight}: WeightsConfig
+): NodeWeight {
+  let totalWeight = 1;
+  for (const tag of tags) {
+    const weight = tagWeights.get(tag);
+    if (weight != null) {
+      totalWeight *= weight;
+    } else {
+      totalWeight *= defaultTagWeight;
+    }
+  }
```

Comment:
This is incorrect. Consider the case where we have defaultTagWeight as 2, no weights on any particular tags, and a like is associated with 2 tags. Then we will incorrectly get a weight of `4` because we multiply by the default weight twice.

I suggest the following code:

```js
const weightForTag = (tagId: TagId) => tagWeights.get(tagId) || 1;
return tags.reduce((acc, tag) => acc * weightForTag(tag), defaultTagWeight)
```
```

Reply:
Your suggestion is backwards, I think.

It should be:
```js
const weightForTag = (tagId: TagId) => either(tagWeights.get(tagId), defaultTagWeight);
  return tags.reduce((acc, tag) => acc * weightForTag(tag), 1);
```

Reply:
> This is incorrect. Consider the case where we have defaultTagWeight as 2, no weights on any particular tags, and a like is associated with 2 tags. Then we will incorrectly get a weight of `4` because we multiply by the default weight twice.

…why? @topocount and I discussed this, and we don’t see why this should
be the case. We interpret `defaultTagWeight` as meaning “any tag whose
weight is not given in `tagWeights` is taken to have this weight”.
Logically, this should mean that you multiply by the default weight
multiple times if there are multiple unweighted tags.

So, if your config has `defaultTagWeight: 3` and explicit tag weights
`A: 2` and `B: 5`, then:

  - a like with no tags has weight 1;
  - a like with tag `A` has weight 2;
  - a like with tags `A` and `B` has weight 10;
  - a like with tags `A`, `B`, and `C` has weight 30;
  - a like with tags `A`, `B`, `C`, and `D` has weight 90.

This is just the natural product semantics.

Is this not how you interpret `defaultTagWeight`? I can’t think of a
sensible interpretation that leads to the semantics that you describe.
If we pick a different interpretation, we should add comments to
`discourse/weights.js` to document the semantics.